omnigibson/utils/bddl_utils.py

```diff
@@ -89,9 +127,810 @@ def process_single_condition(condition):
+
+        # Start with the non-sampleable objects as the first sampled set, then infer recursively
+        cur_batch = self._non_sampleable_object_instances
+        while len(cur_batch) > 0:
+            next_batch = set()
+            for condition, _ in self._object_sampling_conditions["kinematic"]:
+                if condition.body[1] in cur_batch:
+                    next_batch.add(condition.body[0])
+            cur_batch = next_batch
+            self._object_sampling_orders["kinematic"].append(cur_batch)
+        # pop final value since it's an empty set
+        self._object_sampling_orders["kinematic"].pop(-1)
+
+        # Now parse particles -- this time, in reverse order, starting from the final kinematic group
+        obj_insts_to_system = {cond[0]: cond[1] for cond in sampling_groups["particle"]}
+        sampled_particle_entities = set()
+        for batch in reversed(self._object_sampling_orders["kinematic"] + [self._non_sampleable_object_instances]):
+            cur_batch = set()
+            for obj_inst in batch:
+                if obj_inst in obj_insts_to_system:
+                    sampled_particle_entities.add(obj_insts_to_system.pop(obj_inst))
+                    cur_batch.add(obj_inst)
+            self._object_sampling_orders["particle"].append(cur_batch)
+        # Finally, make group with all remaining object instances requiring particle sampling
+        self._object_sampling_orders["particle"].append(set(obj_insts_to_system.keys()))
+
+        # Finally, parse unaries -- this is simply unordered, since it is assumed that unary predicates do not
+        # affect each other
+        self._object_sampling_orders["unary"] = {cond[0] for cond in sampling_groups["unary"]}
+
+        # Aggregate future objects
+        self._future_obj_instances = {cond[0] for cond in sampling_groups["future"]}
+        nonparticle_entities = set(self._object_scope.keys()) - self._substance_instances
+
+        # Sanity check kinematic objects -- any non-system must be kinematically sampled
+        remaining_kinematic_entities = nonparticle_entities - self._future_obj_instances - \
+            self._non_sampleable_object_instances - set.union(*(self._object_sampling_orders["kinematic"] + [set()]))
+        if len(remaining_kinematic_entities) != 0:
+            return f"Some objects do not have any kinematic condition defined for them in the initial conditions: " \
+                   f"{', '.join(remaining_kinematic_entities)}"
+
+        # Sanity check particle systems -- any non-future system must be sample as part of particle groups
+        remaining_particle_entities = self._substance_instances - self._future_obj_instances - sampled_particle_entities
+        if len(remaining_particle_entities) != 0:
+            return f"Some systems do not have any particle condition defined for them in the initial conditions: " \
+                   f"{', '.join(remaining_particle_entities)}"
```

Comment:
how is `insource` being handled now?

Reply:
I don't think it's being handled currently -- it's one of the new predicates right?

Reply:
done!